### ISSUE TITLE
Repeated calls to generator and discriminator's forward in GAN tutorial

### ISSUE THREAD

**Issue by jhauret** [Lightning-AI/tutorials#306]
### 📚 Documentation

In the [training_step of the GAN(L.LightningModule)](https://lightning.ai/docs/pytorch/2.1.0/notebooks/lightning_examples/basic-gan.html#:~:text=def%20training_step(self%2C%20batch)%3A) the generator and discriminator forward are called several times on the same input. Obviously this slows down the training because more computation is required. I wonder if we could just reuse the results of the first call. After all, the `toggle_optimiser`/`untoggle_optimiser` functions should make it safe, right?

For the generator:
- First call: to log images `self.generated_imgs = self(z)`
- Second call:  Inside the generator optimization `self.discriminator(self(z))` 
- Third call:  Inside the discriminator optimization `self.discriminator(self(z).detach())`

For the discriminator:
- First call: Inside the generator optimization `self.discriminator(self(z))` 
- Second call: Inside the discriminator optimization `self.discriminator(self(z).detach())`

cc @borda

**Comment by jhauret** [comment]
Still on the [basic GAN tutorial](https://github.com/Lightning-AI/tutorials/blob/main/lightning_examples/basic-gan/gan.py), I spotted a few more track of improvements:
- [on_validation_epoch_end](https://github.com/Lightning-AI/tutorials/blob/37e07988f981913721a613ac3ab40f53dcf6c9b2/lightning_examples/basic-gan/gan.py#L244) is ignored because `validation_step` is not defined.
-  After training, the quality of the samples is still very poor. I understand that the code is just an entry point for newcomers, but this poor performance makes the user doubtful about the tutorial. We could easily improve this by using a small convolutional architecture for both the generator and the discriminator without complicating the code. (_e.g._  [Generator](https://gist.github.com/SimpleSchwarz/299d1429d6b1132036b43ce6ff1a5a71), [Discriminator](https://gist.github.com/SimpleSchwarz/becfbc58f74139bd8b4dc5f1be85861f) )
- [add_image](https://github.com/Lightning-AI/tutorials/blob/37e07988f981913721a613ac3ab40f53dcf6c9b2/lightning_examples/basic-gan/gan.py#L196) called with the constant argument `global_step=0` is overwriting the results of previous epochs.
- The repeated calls to the generator and discriminators could be simplified to improve training speed.

I am willing to implement and submit a PR if you find this helpful :smiley:

**Comment by Borda** [comment]
Hello @jhauret, I apologize for the late reply, it was not in good shape until a few days ago, so would you be interested in sending PR with the fix? :rabbit:

**Comment by jhauret** [comment]
Hi @Borda, I hope you're feeling better now. Yes, I'll do it the first week of August!
